### PR TITLE
App

### PR DESCRIPTION
**Pull request by onlyskin**
Implement app class to run a basic single TTT game

### CODE REVIEW

File: lib/app.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+require_relative 'board.rb'
```
Comment (fabientownsend) [inline lib/app.rb:1]: I would avoid relative path as much as you can. You can take a look here to see how you can do that: http://fabientownsend.github.io/apprenticeship/2016/10/05/binaire2.html

File: lib/app.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+class App
+    def initialize(input=$stdin, output=$stdout, board=Board.new)
```
Comment (fabientownsend) [inline lib/app.rb:4]: In order to stay consistent, it could be nicer to keep spaces around `=` as you did in the line below

File: lib/app.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        @output = output
+        @input = input
+        @board = board
+    end
+
+    def run
+        output(@board.to_s)
+        while @board.available_moves.length > 0
+            break if @board.game_over?
+            move = input.to_i
```
Comment (fabientownsend) [inline lib/app.rb:14]: What would happen here if the input is a string?

File: lib/app.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+class App
+    def initialize(input=$stdin, output=$stdout, board=Board.new)
+        @output = output
+        @input = input
+        @board = board
+    end
+
+    def run
+        output(@board.to_s)
+        while @board.available_moves.length > 0
```
Comment (fabientownsend) [inline lib/app.rb:12]: It could be good to abstract this notion of length into your class Board

``` ruby
def available_moves?
   @board.available_moves.length > 0
end
```
This look like the feature envy code "smells"
https://sourcemaking.com/refactoring/smells/feature-envy

File: spec/app_spec.rb [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                input = StringIO.new("7\n4\n6\n2\n8\n9\n3\nthis_doesnt_get_read\n")
+                output = StringIO.new()
+                board_spy = BoardSpy.new()
+                board_spy.play(1)
+                board_spy.play(5)
+                app = App.new(input, output, board_spy)
+                app.run()
+                expect(board_spy.play_called).to be true
+                expect(board_spy.play_called_with).to eq([1, 5, 7, 4, 6, 2, 8, 9, 3])
+            end
```
Comment (fabientownsend) [inline spec/app_spec.rb:59]: Would be easier to read with extra line between methods

**Review by fabientownsend — COMMENTED** [review]
👍 
ps: I think Ruby convention is two spaces

File: spec/app_spec.rb (hunk 5 above)
Comment (RabeaGleissner) [reply spec/app_spec.rb:59]: Agree! Definitely think there should be empty lines between methods 😄

File: lib/board.rb [hunk 6]
```diff
@@ -56,6 +56,17 @@ def play(move)
         @cells[move-1] = current_marker
     end
 
+    def to_s()
```
Comment (RabeaGleissner) [inline lib/board.rb:59]: I would leave out the `()` if the method has no parameters

Comment (RabeaGleissner) [reply lib/board.rb:59]: `to_s` is an existing Ruby method. Did you consciously make the choice to overwrite it here? Why?

File: lib/board.rb [hunk 7]
```diff
@@ -56,6 +56,17 @@ def play(move)
...
+        (0..8).each do |i|
+            result << @cells[i]
+            if i % 3 == 2
+                result << "\n"
+            end
+        end
+        result
+    end
+
     def _full?
```
Comment (RabeaGleissner) [inline lib/board.rb:70]: I haven't seen Ruby methods starting with an underscore. Why did you add it? I'd remove it.

File: spec/app_spec.rb [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            it 'stops when board.game_over? is true' do
+                input = StringIO.new("1\n2\n4\n5\n7\n")
+                output = StringIO.new()
+                board_spy = BoardSpy.new()
+                app = App.new(input, output, board_spy)
+                app.run()
+                expect(board_spy.play_called_with).to eq([1, 2, 4, 5, 7])
+            end
+            it 'calls to_s on board_spy six times' do
```
Comment (RabeaGleissner) [inline spec/app_spec.rb:68]: I'm not sure about this test name. I think it's too technical and too tied to implementation details.  Why do you need to call to_s six times? Maybe something like "turns board into printable string for each move made" would be better?

File: spec/board_spec.rb [hunk 9]
```diff
@@ -173,16 +153,26 @@
...
+        context 'to_s called' do
+            it 'returns string representation' do
+                board = make_board_and_play_moves([7, 8, 5, 6])
+                expect(board.to_s).to eq("---\n-XO\nXO-\n")
+            end
+        end
+    end
+end
+
+def make_board_and_play_moves(moves)
```
Comment (RabeaGleissner) [inline spec/board_spec.rb:172]: Not sure about this method name. It's very literal as to what the method does. Then you almost need to say `make_board_and_play_moves_and_return_board_afterwards` 😹 
so maybe just call it `board_with_moves`

**Review by RabeaGleissner — COMMENTED** [review]
So far so good! Looking forward to more Ruby :)
